File: Aran/RNAV/Libs/Panda/Geometry/C#/Point.cs
using System;
using System.Collections.Generic;
using System.Text;
using ARAN.Common;
using ARAN.Contracts.Registry;

namespace ARAN.GeometryClasses
{

	public class Point : Geometry
	{
		public Point()
			: base()
		{
			SetEmpty();
		}

		public Point(double initX, double initY)
			: base()
		{
			SetEmpty();
			_x = initX;
			_y = initY;
		}

		public Point(double initX, double initY, double initZ)
			: base()
		{
			SetEmpty();
			_x = initX;
			_y = initY;
			_z = initZ;
		}
		public Point(Point point)
			: base()
		{
			Assign(point);
		}


		public void SetCoords(double x, double y)
		{
			_isEmpty = false;
			_x = x;
			_y = y;
		}

		public void SetCoords(double x, double y, double z)
		{
			_isEmpty = false;
			_x = x;
			_y = y;
			_z = z;
		}

		public void SetEmpty()
		{
			_isEmpty = true;
			_x = Double.NaN;
			_y = Double.NaN;
			_z = 0.0;
			_t = 0.0;
			_m = 0.0;
		}

		public double X
		{
			get { return _x; }
			set
			{
				_x = value;
				_isEmpty = false;
			}
			//_isEmpty = false;

		}

		public double Y
		{
			get { return _y; }
			set
			{
				_y = value;
				_isEmpty = false;
			}
		}

		public double Z
		{
			get { return _z; }
			//_isEmpty = false;
			set
			{
				_z = value;
				_isEmpty = false;
			}
		}

		public double T
		{
			get
			{
				return _t;
				//_isEmpty = false;
			}
			set
			{
				_t = value;
				_isEmpty = false;
			}
		}

		public double M
		{
			get { return _m; }//_isEmpty = false;
			set
			{
				_m = value;
				_isEmpty = false;
			}
		}

		public bool IsEmpty()
		{
			return _isEmpty;
		}

		public override Object Clone()
		{
			Point cln = new Point();
			cln.Assign(this);
			return cln;
		}

		public override void Assign(PandaItem source)
		{
			Point src = (Point)source;
			_isEmpty = src._isEmpty;
			_x = src._x;
			_y = src._y;
			_z = src._z;
			_t = src._t;
			_m = src._m;
		}

		public override void Pack(int handle)
		{
			Registry_Contract.PutDouble(handle, _x);
			Registry_Contract.PutDouble(handle, _y);
			Registry_Contract.PutDouble(handle, _z);
			Registry_Contract.PutDouble(handle, _m);
			Registry_Contract.PutDouble(handle, _t);
		}

		public override void UnPack(int handle)
		{
			_x = Registry_Contract.GetDouble(handle);
			_y = Registry_Contract.GetDouble(handle);
			_z = Registry_Contract.GetDouble(handle);
			_m = Registry_Contract.GetDouble(handle);
			_t = Registry_Contract.GetDouble(handle);
		}

		public override GeometryType GetGeometryType()
		{
			return GeometryType.Point;
		}

		private double _x = 0;
		private double _y = 0;
		private double _z = 0;
		private double _t = 0;
		private double _m = 0;
		private bool _isEmpty = true;
	}
}

File: Aran/RNAV/Libs/Panda/Geometry/cpp/src/Point.cpp
#include "../include/Point.h"
#include "../include/GeometryType.h"

namespace Panda
{

	Point::Point() throw ()
	{
		setEmpty ();
	}

	Point::Point (double x, double y) throw ()
	{
		setCoords (x, y);
	}

	void Point::setCoords (double x, double y) throw ()
	{
		setX(x);
		setY(y);
	}

	void Point::setX (double x) throw ()
	{
		_x = x;
	}

	double Point::getX () const throw ()
	{
		return _x;
	}

	void Point::setY(double y) throw ()
	{
		_y = y;
	}

	double Point::getY() const throw ()
	{
		return _y;
	}

	void Point::setZ(double z) throw ()
	{
		_z = z;
	}

	double Point::getZ() const throw ()
	{
		return _z;
	}

	void Point::setM(double m) throw ()
	{
		_m = m;
	}

	double Point::getM() const throw ()
	{
		return _m;
	}

	void Point::setT(double t) throw ()
	{
		_t = t;
	}

	double Point::getT() const throw ()
	{
		return _t;
	}

	bool Point::isEmpty() const throw ()
	{
		return  (isNaN(_x) || isNaN(_y));
	}

	void Point::setEmpty() throw ()
	{
		_x = NaN;
		_y = NaN;
		_z = NaN;
		_t = NaN;
		_m = NaN;
	}

	void Point::assign (const Geometry& geometry) throw (std::bad_cast)
	{
		const Point& src = dynamic_cast <const Point&> (geometry);
		_x = src._x;
		_y = src._y;
		_z = src._z;
		_m = src._m;
		_t = src._t;
	}

	Geometry* Point::clone() const throw (std::bad_cast)
	{
		Point* p = new Point();
        
        try
        {
		    p->assign (*this);
        }
        catch (std::bad_cast e)
        {
            delete p;
            throw e;
        }
        
		return p;
	}

	void Point::pack (Handle handle) const throw (Registry::Exception)
	{
		Registry::putDouble (handle, _x);
		Registry::putDouble (handle, _y);
		Registry::putDouble (handle, _z);
		Registry::putDouble (handle, _m);
		Registry::putDouble (handle, _t);
	}

	void Point::unpack (Handle handle) throw (Registry::Exception)
	{
		Registry::getDouble (handle, _x);
		Registry::getDouble (handle, _y);
		Registry::getDouble (handle, _z);
		Registry::getDouble (handle, _m);
		Registry::getDouble (handle, _t);
	}

	int Point::geometryType () const throw ()
	{
		return GeometryType::Point;
	}
}
